### DISCUSSION TITLE
Implementing minimal `hb_font_funcs_t` for working with shaper

### DISCUSSION THREAD

**Discussion by nisargjhaveri** [harfbuzz/harfbuzz#4798] [Q&A] [answered]
I'm trying to implement a simple text renderer that can support variety of text (at least indic languages if not more) for an embedded system with very limited resources. I can get harfbuzz running. I don't want to run freetype with entire ttf files, for size and perf/speed considerations.

Here is what I'm thinking:
- Use hb-subset to get minimal number of glyphs needed for a given text or language/script from a font.
- Have those glyphs pre-rendered (bitmap) in custom format, along with offset/advances information.
- Run harfbuzz without in-built font functions
- Provide custom font-functions with `hb_font_set_funcs` for the shaper.

Here is what I'm struggling with:
1. How does Harfbuzz query or get glyphs for clusters from the font API? `hb_font_get_nominal_glyph_func_t` and `hb_font_get_nominal_glyphs_func_t` seems to be only getting the glyphs for single codepoints and not a sequence of codepoints (though I also don't fully understand the function of different `hb_font_get_nominal_glyphs_func_t`)? How does it map codepoint for half or combined characters in indic languages?
1. What all can I safely not implement from the `font_funcs`? For e.g., I believe I won't need to implement the callbacks related to draw/paint if Im only going to use shaper. Can I not implement the `get_glyph_from_name` and `get_glyph_name` apis? Anything else I can avoid implementing to minimise required embedded data?

Any pointers would help! Thanks!

**Comment by behdad** [comment, marked as answer]
I highly recommend using the built-in font-funcs. Have you read `CONFIG.md` re how to make the library smaller?

> Here is what I'm struggling with:
> 
> 1. How does Harfbuzz query or get glyphs for clusters from the font API? `hb_font_get_nominal_glyph_func_t` and `hb_font_get_nominal_glyphs_func_t` seems to be only getting the glyphs for single codepoints and not a sequence of codepoints (though I also don't fully understand the function of different `hb_font_get_nominal_glyphs_func_t`)? How does it map codepoint for half or combined characters in indic languages?

Those happen later by using the data in the GSUB table. There's no font-func involved.

> 2. What all can I safely not implement from the `font_funcs`? For e.g., I believe I won't need to implement the callbacks related to draw/paint if Im only going to use shaper. Can I not implement the `get_glyph_from_name` and `get_glyph_name` apis? Anything else I can avoid implementing to minimise required embedded data?

You need the `get_nominal_glyph` and `get_h_advance` functions as a minimum.
